File: android/app/src/main/java/be/peter/twizypitpro/Tuning.java
package be.peter.twizypitpro;
import org.json.*;
import java.util.*;

/** Register targets, adapted from M. Balzer's OVMS (2017, MIT).
 * See reference/rt_sevcon_tuning.cpp and the included OVMS license.
 * This compiler does not transmit and does not qualify hardware. */
final class Tuning {
    private static final long[] C80={80,7250,900,400,800,1250,89,8050,550,55000,57000,0,450000,450,540,12182,13000,4608};
    private static final long[] C45={45,5814,1307,686,1386,2686,56,7200,900,32500,33000,500,270000,290,330,7050,7650,2688};
    private static long scale(long d,long base,long v,long min,long max){return v==base?d:Math.max(min,Math.min(max,d*v/base));}
    private static void add(JSONArray rows,int index,int sub,Long raw,String keys,int width,boolean signed) {
        if(raw!=null&&(raw<(signed?-(1L<<(width*8-1)):0)||raw>=(1L<<(width*8-(signed?1:0)))))throw new IllegalArgumentException("Registerwaarde buiten datatype");
        JSONObject r=new JSONObject();Model.put(r,"index",index);Model.put(r,"sub",sub);Model.put(r,"address",String.format(Locale.ROOT,"%04X:%02X",index,sub));Model.put(r,"width",width);Model.put(r,"signed",signed);Model.put(r,"raw",raw==null?JSONObject.NULL:raw);Model.put(r,"keys",new JSONArray(Arrays.asList(keys.split(","))));Model.put(r,"key",keys.replace(","," / "));rows.put(r);
    }
    private static void add(JSONArray rows,int index,int sub,long raw,String keys){add(rows,index,sub,raw,keys,2,false);}
    static JSONArray targets(Model model,JSONObject v,String type,Long flags) {
        if(!type.equals("80")&&!type.equals("45"))throw new IllegalArgumentException("Onbekende aandrijflijn");
        model.validate(v,type);boolean t45=type.equals("45");long[] c=t45?C45:C80;JSONArray rows=new JSONArray();
        String[] simple={"drive","neutral","brake","ramp_start","ramp_accel","ramp_decel","ramp_neutral","ramp_brake","rampl_accel","rampl_decel","smooth"};
        int[] indices={0x2920,0x2920,0x2920,0x291c,0x2920,0x2920,0x2920,0x2920,0x2920,0x2920,0x290a},subs={1,3,4,2,7,11,13,14,15,16,3};
        long[] def={1000,t45?209:182,t45?209:182,t45?300:400,t45?2083:2500,2000,4000,4000,6000,6000,800},base={100,t45?21:18,t45?21:18,t45?30:40,t45?21:25,20,40,40,30,30,70},min={0,0,0,10,10,10,10,10,0,0,0},max={1000,1000,1000,10000,10000,10000,10000,10000,20000,20000,1000};
        for(int i=0;i<simple.length;i++)add(rows,indices[i],subs[i],scale(def[i],base[i],v.optLong(simple[i]),min[i],max[i]),simple[i],2,i==10);
        add(rows,0x290a,1,1+v.optLong("smooth")/10,"smooth",1,false);
        long rpm=scale(c[1],c[0],v.optLong("speed"),400,65535),warning=scale(c[7],c[6],v.optLong("warn"),400,65535);
        if(warning<c[8])throw new IllegalArgumentException("Waarschuwingssnelheid te laag voor hysterese");
        add(rows,0x3813,0x34,warning,"warn");add(rows,0x3813,0x3c,warning-c[8],"warn");
        add(rows,0x2920,5,rpm,"speed");add(rows,0x2920,6,Math.min(rpm,c[2]),"speed");
        int[] speedSubs={0x33,0x35,0x3b,0x2d};long[] offsets={c[3],c[4],c[5],c[5]+1500};
        for(int i=0;i<4;i++)add(rows,0x3813,speedSubs[i],rpm+offsets[i],"speed");
        add(rows,0x4624,0,rpm+c[5]+2500,"speed");
        add(rows,0x4641,2,scale(c[13],100,v.optLong("current"),0,c[14]),"current");
        add(rows,0x6075,0,scale(c[12],100,v.optLong("current"),0,c[14]*1000),"current",4,false);
        long torque=scale(c[9],100,v.optLong("torque"),10000,200000),low=scale(c[15],100,v.optLong("power_low"),500,200000),high=scale(c[16],100,v.optLong("power_high"),500,200000);
        add(rows,0x6076,0,torque+c[11],"torque",4,false);add(rows,0x2916,1,v.optInt("torque")==100?c[10]:torque+c[11],"torque",4,false);
        add(rows,0x3813,0x23,v.optInt("power_low")==100&&v.optInt("power_high")==100?c[17]:(long)(Math.max(low,high)*.353),"power_low,power_high",2,true);
        long[] fmap=t45?new long[]{480,8192,576,8960}:new long[]{964,9728,1122,9984};
        long[] pmap;
        if(rpm==c[1]&&torque==c[9]&&low==c[15]&&high==c[16])pmap=t45?new long[]{520,0,520,2050,437,2500,363,3000,314,3500,279,4000,247,4500,226,5000,195,6000}:new long[]{880,0,880,2115,659,2700,608,3000,516,3500,421,4500,360,5500,307,6500,273,7250};
        else {
            long rpm2=Math.min(low*9549/torque,rpm),trq=(torque*16+500)/1000;
            if(1000/9.549*10000*10000*(v.optDouble("current")/100)/high<rpm||rpm2<=0)throw new IllegalArgumentException("Niet ondersteunde breakdown-configuratie");
            if(trq>fmap[2])fmap=t45?new long[]{656,9600,1328,11901}:new long[]{1122,10089,2240,11901};
            pmap=new long[18];pmap[0]=trq;long drpm=(rpm-rpm2)/21,dpwr=(high-low)/21;int[] fib={0,1,2,3,5,8,13,21};
            for(int i=0;i<8;i++){long pr=rpm2+fib[i]*drpm,pp=low+fib[i]*dpwr;pmap[2+i*2]=(((pp*9549+(pr>>1))/pr)*16+500)/1000;pmap[3+i*2]=pr;}
        }
        String powerkeys="speed,torque,power_low,power_high,current";
        for(int i=0;i<18;i++)add(rows,0x4611,i+1,pmap[i],powerkeys);
        for(int i=0;i<4;i++)add(rows,0x4610,15+i,fmap[i],powerkeys);
        long[] mapspeed=t45?new long[]{4357,5083,6535,8714}:new long[]{3000,3500,4500,6000};int[] speeds={33,39,50,66},bases={0x24,0x1b,7};String[] letters={"D","N","B"};
        for(int m=0;m<3;m++){long previous=-1;for(int i=0;i<4;i++){
            String levelkey=letters[m]+"_level"+(i+1),speedkey=letters[m]+"_speed"+(i+1);
            add(rows,0x3813,bases[m]+i*2,scale(32767,100,v.optLong(levelkey),0,32767),levelkey);
            long point=scale(mapspeed[i],speeds[i],v.optLong(speedkey),0,65535);
            if(point<=previous)throw new IllegalArgumentException("Omgerekende kaartpunten lopen niet op");previous=point;
            add(rows,0x3813,bases[m]+i*2+1,point,speedkey);
        }}
        add(rows,0x3813,5,scale(1024,100,v.optLong("brakelight_off"),64,1024),"brakelight_off");add(rows,0x3813,6,scale(1024,100,v.optLong("brakelight_on"),64,1024),"brakelight_on");
        if(flags!=null&&(flags<0||flags>65535))throw new IllegalArgumentException("Ongeldige control flags");
        boolean enabled=v.optInt("brakelight_on")!=100||v.optInt("brakelight_off")!=100;
        add(rows,0x2910,1,flags==null?null:(enabled?flags|0x2000:flags&~0x2000),"brakelight_on,brakelight_off",2,false);
        return rows;
    }
    static JSONArray compare(Model model,JSONObject values,String type,JSONObject scan){
        Map<String,JSONObject> previous=new HashMap<>();
        if(scan!=null&&scan.optString("mode").equals("live-usb")){
            JSONArray rs=scan.optJSONArray("registers");if(rs!=null)for(int i=0;i<rs.length();i++){JSONObject r=rs.optJSONObject(i);if(r!=null&&!r.isNull("raw")&&!r.has("error"))previous.put(r.optString("address"),r);}
        }
        JSONObject control=previous.get("2910:01");Long flags=control!=null&&control.optInt("width")==2?control.optLong("raw"):null;
        JSONArray rows=targets(model,values,type,flags);
        for(int i=0;i<rows.length();i++){JSONObject r=rows.optJSONObject(i),old=previous.get(r.optString("address"));boolean known=old!=null&&old.optInt("width")==r.optInt("width");Model.put(r,"before",known?old.optLong("raw"):JSONObject.NULL);Model.put(r,"verified_width",known);Model.put(r,"changed",!known||r.isNull("raw")||r.optLong("raw")!=old.optLong("raw"));}
        return rows;
    }
}

File: pit/tuning.py
"""Complete register target compiler; no bus writes.

Adapted from Michael Balzer's OVMS rt_sevcon_tuning.cpp (2017), MIT.
The original notice is retained in reference/rt_sevcon_tuning.cpp.
Only the standard 45/80 drivetrain and default OVMS breakdown calculation
are modeled. A target set is not a qualified or ordered live transaction.
"""
from .catalog import defaults, parameters, scale, sdo_candidates, validate


CONFIG = {
    "80": dict(kph=80, rpm=7250, reverse=900, offsets=(400, 800, 1250), warn=89,
               warnrpm=8050, hysteresis=550, torque=55000, rated=57000, delta=0,
               current=450000, stator=450, boost=540, low=12182, high=13000,
               motorpower=4608, fmap=(964,9728,1122,9984), extended=(1122,10089,2240,11901),
               pmap=(880,0,880,2115,659,2700,608,3000,516,3500,421,4500,360,5500,307,6500,273,7250),
               mapspeed=(3000,3500,4500,6000)),
    "45": dict(kph=45, rpm=5814, reverse=1307, offsets=(686,1386,2686), warn=56,
               warnrpm=7200, hysteresis=900, torque=32500, rated=33000, delta=500,
               current=270000, stator=290, boost=330, low=7050, high=7650,
               motorpower=2688, fmap=(480,8192,576,8960), extended=(656,9600,1328,11901),
               pmap=(520,0,520,2050,437,2500,363,3000,314,3500,279,4000,247,4500,226,5000,195,6000),
               mapspeed=(4357,5083,6535,8714)),
}


def _trunc_div(value, divisor):
    return (abs(value) // divisor) * (-1 if value < 0 else 1)


def targets(values, model, control_flags=None):
    """Map all 43 fields, including dependent limits, PMAP/FMAP and masked flags.

    Missing control_flags stays unresolved: never substitute a factory value
    for a read-modify-write register. Rows describe targets, not execution order.
    """
    if model not in CONFIG:
        raise ValueError("Onbekende aandrijflijn; geen registervertaling.")
    v, c = validate(values, model), CONFIG[model]
    rows = []
    def add(index, sub, raw, keys, width=2, signed=False, **extra):
        if raw is not None and not (-(1 << (width*8-1)) if signed else 0) <= raw < (1 << (width*8-(1 if signed else 0))):
            raise ValueError(f"{index:04X}:{sub:02X}: berekende waarde past niet in datatype.")
        rows.append(dict(index=index, sub=sub, address=f"{index:04X}:{sub:02X}",
                         width=width, signed=signed, raw=raw, keys=list(keys), key=" / ".join(keys), **extra))
    for r in sdo_candidates(v, model):
        add(r["index"], r["sub"], r["raw"], [r["key"]], r["width"], r["address"] == "290A:03")
    rpm = scale(c["rpm"], c["kph"], v["speed"], 400, 65535)
    warning = scale(c["warnrpm"], c["warn"], v["warn"], 400, 65535)
    # Some editor ranges were broad enough to underflow this subtraction.
    if warning < c["hysteresis"]:
        raise ValueError("Waarschuwingssnelheid te laag voor de hysterese van dit model.")
    add(0x3813, 0x34, warning, ["warn"])
    add(0x3813, 0x3c, warning-c["hysteresis"], ["warn"])
    add(0x2920, 5, rpm, ["speed"])
    add(0x2920, 6, min(rpm, c["reverse"]), ["speed"])
    for sub, offset in zip((0x33,0x35,0x3b,0x2d), (*c["offsets"], c["offsets"][2]+1500)):
        add(0x3813, sub, rpm+offset, ["speed"])
    add(0x4624, 0, rpm+c["offsets"][2]+2500, ["speed"])
    add(0x4641, 2, scale(c["stator"],100,v["current"],0,c["boost"]), ["current"])
    add(0x6075, 0, scale(c["current"],100,v["current"],0,c["boost"]*1000), ["current"], 4)
    torque = scale(c["torque"],100,v["torque"],10000,200000)
    low = scale(c["low"],100,v["power_low"],500,200000)
    high = scale(c["high"],100,v["power_high"],500,200000)
    add(0x6076,0,torque+c["delta"],["torque"],4)
    add(0x2916,1,c["rated"] if v["torque"]==100 else torque+c["delta"],["torque"],4)
    add(0x3813,0x23,c["motorpower"] if v["power_low"]==v["power_high"]==100 else int(max(low,high)*.353),["power_low","power_high"],signed=True)
    powerkeys = ["speed","torque","power_low","power_high","current"]
    if (rpm,torque,low,high)==(c["rpm"],c["torque"],c["low"],c["high"]):
        pmap, fmap = c["pmap"], c["fmap"]
    else:
        rpm2 = min(low*9549//torque,rpm)
        # Default OVMS disables custom breakdown parameters. For the bounded
        # editor domain region 3 must remain above the requested maximum RPM.
        rpm3 = 1000/9.549*10000*10000*(v["current"]/100)/high
        if rpm3 < rpm or rpm2 <= 0:
            raise ValueError("Vermogenskaart vereist een nog niet ondersteunde breakdown-configuratie.")
        trq = (torque*16+500)//1000
        fmap = c["extended"] if trq > c["fmap"][2] else c["fmap"]
        pmap = [trq,0]
        drpm, dpwr = (rpm-rpm2)//21, _trunc_div(high-low,21)
        for fib in (0,1,2,3,5,8,13,21):
            point_rpm, point_power = rpm2+fib*drpm, low+fib*dpwr
            point_torque = (((point_power*9549+(point_rpm>>1))//point_rpm)*16+500)//1000
            pmap.extend((point_torque,point_rpm))
    for sub, raw in enumerate(pmap,1): add(0x4611,sub,raw,powerkeys)
    for sub, raw in enumerate(fmap,0x0f): add(0x4610,sub,raw,powerkeys)
    for letter, base in (("D",0x24),("N",0x1b),("B",7)):
        for n, (refspeed, kph) in enumerate(zip(c["mapspeed"],(33,39,50,66)),1):
            speedkey, levelkey = f"{letter}_speed{n}", f"{letter}_level{n}"
            add(0x3813,base+2*(n-1),scale(32767,100,v[levelkey],0,32767),[levelkey], map=letter, point=n, component="level")
            add(0x3813,base+2*(n-1)+1,scale(refspeed,kph,v[speedkey],0,65535),[speedkey], map=letter, point=n, component="speed")
    # Strict editor ordering alone does not ensure ordering after per-point scaling.
    for letter in "DNB":
        speeds=[r["raw"] for r in rows if r.get("map")==letter and r.get("component")=="speed"]
        if any(a>=b for a,b in zip(speeds,speeds[1:])):
            raise ValueError(f"Koppelkaart {letter}: omgerekende toerentalpunten lopen niet strikt op.")
    add(0x3813,5,scale(1024,100,v["brakelight_off"],64,1024),["brakelight_off"])
    add(0x3813,6,scale(1024,100,v["brakelight_on"],64,1024),["brakelight_on"])
    enabled = v["brakelight_on"]!=100 or v["brakelight_off"]!=100
    if control_flags is not None and (type(control_flags) is not int or not 0<=control_flags<=65535):
        raise ValueError("Ongeldige uitgelezen control flags.")
    flags = None if control_flags is None else (control_flags|0x2000 if enabled else control_flags&~0x2000)
    add(0x2910,1,flags,["brakelight_on","brakelight_off"],mask=0x2000,set_bits=0x2000 if enabled else 0)
    return rows


def register_inventory(model):
    return [{k:v for k,v in row.items() if k not in ("raw","set_bits")} for row in targets(defaults(model),model)]


def compare_snapshot(values, model, registers):
    """Bind a target set to an actually read snapshot; don't guess unread values."""
    by_address={r["address"]:r for r in registers if r.get("source")=="live" and r.get("raw") is not None and not r.get("error")}
    control=by_address.get("2910:01",{})
    flags=control.get("raw") if control.get("width")==2 else None
    result=[]
    for row in targets(values,model,flags):
        old=by_address.get(row["address"])
        row=dict(row, before=old["raw"] if old and old.get("width")==row["width"] else None)
        row["verified_width"]=bool(old and old.get("width")==row["width"])
        row["changed"]=row["before"] is None or row["raw"]!=row["before"]
        result.append(row)
    return result


def map_update_order(before, after):
    """Find an order that preserves strictly increasing RPM points at every step."""
    if len(before)!=4 or len(after)!=4 or any(a>=b for seq in (before,after) for a,b in zip(seq,seq[1:])):
        raise ValueError("Ongeldige koppelkaart; geen veilige updatevolgorde.")
    current=list(before);todo={i for i in range(4) if before[i]!=after[i]};order=[]
    while todo:
        candidates=[i for i in sorted(todo) if (i==0 or current[i-1]<after[i]) and (i==3 or after[i]<current[i+1])]
        if not candidates: raise ValueError("Geen begrensde kaartupdatevolgorde gevonden.")
        i=candidates[0];current[i]=after[i];todo.remove(i);order.append(i)
    return order
